utils/frame/try-runtime/cli/src/lib.rs

```diff
@@ -53,169 +86,317 @@ pub struct TryRuntimeCmd {
-		);
-
-		let ext = {
-			use remote_externalities::{Builder, Mode, SnapshotConfig, OfflineConfig, OnlineConfig};
-			let builder = match &self.state {
-				State::Snap { snapshot_path } => {
-					Builder::<B>::new().mode(Mode::Offline(OfflineConfig {
-						state_snapshot: SnapshotConfig::new(snapshot_path),
-					}))
-				},
-				State::Live {
-					url,
-					snapshot_path,
-					block_at,
-					modules
-				} => Builder::<B>::new().mode(Mode::Online(OnlineConfig {
-					transport: url.to_owned().into(),
-					state_snapshot: snapshot_path.as_ref().map(SnapshotConfig::new),
-					modules: modules.to_owned().unwrap_or_default(),
-					at: block_at.as_ref()
-						.map(|b| b.parse().map_err(|e| format!("Could not parse hash: {:?}", e))).transpose()?,
-					..Default::default()
-				})),
-			};
-
-			// inject the code into this ext.
-			builder.inject(&[(code_key, code)]).build().await?
-		};
-
-		let encoded_result = StateMachine::<_, _, NumberFor<B>, _>::new(
-			&ext.backend,
-			None,
-			&mut changes,
-			&executor,
-			"TryRuntime_on_runtime_upgrade",
-			&[],
-			ext.extensions,
-			&sp_state_machine::backend::BackendRuntimeCode::new(&ext.backend)
-				.runtime_code()?,
-			sp_core::testing::TaskExecutor::new(),
-		)
-		.execute(execution.into())
-		.map_err(|e| format!("failed to execute 'TryRuntime_on_runtime_upgrade' due to {:?}", e))?;
-
-		let (weight, total_weight) = <(u64, u64) as Decode>::decode(&mut &*encoded_result)
-			.map_err(|e| format!("failed to decode output due to {:?}", e))?;
-		log::info!(
-			"try-runtime executed without errors. Consumed weight = {}, total weight = {} ({})",
-			weight,
-			total_weight,
-			weight as f64 / total_weight as f64
-		);
-
-		Ok(())
+		match &self.command {
+			Command::OnRuntimeUpgrade(ref cmd) => {
+				on_runtime_upgrade::<B, ExecDispatch>(self.shared.clone(), cmd.clone(), config).await
+			}
+			Command::OffchainWorker(cmd) => {
+				offchain_worker::<B, ExecDispatch>(self.shared.clone(), cmd.clone(), config).await
+			}
+		}
 	}
 }
 
 impl CliConfiguration for TryRuntimeCmd {
 	fn shared_params(&self) -> &sc_cli::SharedParams {
-		&self.shared_params
+		&self.shared.shared_params
 	}
 
 	fn chain_id(&self, _is_dev: bool) -> sc_cli::Result<String> {
-		Ok(match self.shared_params.chain {
+		Ok(match self.shared.shared_params.chain {
 			Some(ref chain) => chain.clone(),
 			None => "dev".into(),
 		})
 	}
 }
+
+/// Extract `:code` from given chain spec and return as `StorageData` along with the corresponding
+/// `StorageKey`.
+fn extract_code(spec: Box<dyn ChainSpec>) -> sc_cli::Result<(StorageKey, StorageData)> {
+	let genesis_storage = spec.build_storage()?;
+	let code = StorageData(
+		genesis_storage
+			.top
+			.get(well_known_keys::CODE)
+			.expect("code key must exist in genesis storage; qed")
+			.to_vec(),
+	);
+	let code_key = StorageKey(well_known_keys::CODE.to_vec());
+
+	Ok((code_key, code))
+}
+
+// Utils for parsing user input
+mod parse {
```

Comment:
Even though the module is small, it's still probably better to break it out into a separate file. Just makes code navigation easier.

Reply:
https://github.com/paritytech/substrate/pull/8966/commits/62a996d93bfedfd198ff03a76b0527db2a02afd1